File: spring-boot/role-management/src/main/java/com/rolemanagement/starter/organisation/OrganisationService.java
package com.rolemanagement.starter.organisation;

import com.rolemanagement.starter.common.exception.ForbiddenException;
import com.rolemanagement.starter.common.exception.NotFoundException;
import com.rolemanagement.starter.organisation.dto.OrganisationRequest;
import com.rolemanagement.starter.organisationInvitation.OrganisationInvitationRepository;
import com.rolemanagement.starter.organisationMemberhsip.OrganisationMembershipService;
import com.rolemanagement.starter.role.Role;
import com.rolemanagement.starter.role.RoleService;
import com.rolemanagement.starter.userTable.UserService;
import com.rolemanagement.starter.userTable.UserTable;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class OrganisationService {
    private final OrganisationRepository organisationRepository;
    private final OrganisationMembershipService organisationMembershipService;
    private final OrganisationInvitationRepository organisationInvitationRepository;
    private final RoleService roleService;
    private final UserService userService;

    public Organisation getById(Long id) {
        return organisationRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Organisation not found: " + id));
    }

    public Organisation create(OrganisationRequest request, String userEmail) {
        Organisation organisation = organisationRepository.save(
                Organisation.builder().name(request.name()).type(request.type()).build()
        );
        Role role = roleService.createDefaultRole(organisation.getId(), "Admin");
        UserTable user = userService.getByEmail(userEmail);
        organisationMembershipService.assignOwner(organisation.getId(), user.getId(), role.getId());
        return organisation;
    }

    public Organisation selectOrganisation(Long organisationId, String userEmail) {
        UserTable user = userService.getByEmail(userEmail);
        if (!organisationMembershipService.userBelongsToOrganisation(userEmail, organisationId)) {
            throw new NotFoundException("User is not a member of the organisation: " + organisationId);
        }

        return getById(organisationId);
    }

    public Organisation update(Long organisationId, OrganisationRequest request, String userEmail) {
        requireOwner(organisationId, userEmail);
        Organisation organisation = getById(organisationId);
        organisation.setName(request.name());
        organisation.setType(request.type());
        return organisationRepository.save(organisation);
    }

    @Transactional
    public void delete(Long organisationId, String userEmail) {
        requireOwner(organisationId, userEmail);
        Organisation organisation = getById(organisationId);
        organisationInvitationRepository.deleteAll(organisationInvitationRepository.findByOrganisationId(organisationId));
        organisationMembershipService.deleteAllByOrganisation(organisationId);
        roleService.deleteAllByOrganisation(organisationId);
        organisationRepository.delete(organisation);
    }

    private void requireOwner(Long organisationId, String userEmail) {
        if (!organisationMembershipService.isOwner(userEmail, organisationId)) {
            throw new ForbiddenException("Only the organisation owner can perform this action");
        }
    }

}

File: nestjs/role-management/src/organisation/organisation.service.ts
import { ConflictException, ForbiddenException, Injectable, NotFoundException } from '@nestjs/common';
import { CreateOrganisationDto } from './dto/create-organisation.dto';
import { UpdateOrganisationDto } from './dto/update-organisation.dto';
import { AddUserToOrganisationDto } from './dto/add-user-to-organisation.dto';
import { InjectRepository } from '@nestjs/typeorm';
import { Organisation } from './entities/organisation.entity';
import { Repository } from 'typeorm';
import { Role } from 'src/role/entities/role.entity';
import { JwtPayload } from 'src/auth/interfaces/jwt-payload.interface';
import { UserTable } from 'src/user-table/entities/user-table.entity';
import { OrganisationMembership } from 'src/organisation-membership/entities/organisation-membership.entity';
import { MembershipInvitation } from 'src/memberhsip-invitation/entities/membership-invitation.entity';
import { EmailService } from 'src/email/email.service';

@Injectable()
export class OrganisationService {
  constructor(
    @InjectRepository(Organisation)
    private readonly organisationRepository: Repository<Organisation>,

    @InjectRepository(Role)
    private readonly roleRepository: Repository<Role>,

    @InjectRepository(UserTable)
    private readonly userRepository: Repository<UserTable>,

    @InjectRepository(OrganisationMembership)
    private readonly membershipRepository: Repository<OrganisationMembership>,

    @InjectRepository(MembershipInvitation)
    private readonly invitationRepository: Repository<MembershipInvitation>,

    private readonly emailService: EmailService,
  ) {}

  private async ensureOwner(organisationId: number, userId: number) {
    const membership = await this.membershipRepository.findOne({
      where: {
        organisation: { id: organisationId },
        user: { id: userId },
      },
      relations: ['organisation'],
    })

    if (!membership || !membership.isOwner) {
      throw new ForbiddenException('Only the organisation owner can perform this action')
    }

    return membership.organisation
  }

  async create(createOrganisationDto: CreateOrganisationDto, user: JwtPayload) {
    const organisation = this.organisationRepository.create(createOrganisationDto);
    await this.organisationRepository.save(organisation)
    const savedRole = this.roleRepository.create({
      name: 'Admin',
      isSystemRole: true,
      organisation,
    })
    await this.roleRepository.save(savedRole)
    const currUser = await this.userRepository.findOne({ where: {
      email: user.email
    }})
    if (!currUser) {
      throw new NotFoundException('User not found')
    }

    const newMembership = this.membershipRepository.create({
      user: currUser,
      organisation,
      role: savedRole,
      isOwner: true,
    })
    await this.membershipRepository.save(newMembership)
    return this.organisationRepository.save(organisation);
  }

  async selectOrganisation(id: number, user: JwtPayload) {
    const membership = await this.membershipRepository.findOne({
      where: {
        organisation: { id },
        user: { id: user.sub },
      },
      relations: ['organisation'],
    })

    if (!membership) {
      throw new ForbiddenException('You are not a member of this organisation')
    }

    return membership.organisation
  }

  async addUserToOrganisation(id: number, organisationId: number, addUserToOrganisationDto: AddUserToOrganisationDto) {
    if (id !== organisationId) {
      throw new ForbiddenException('You can only add members to your currently selected organisation')
    }

    const user = await this.userRepository.findOne({ where: { email: addUserToOrganisationDto.email } })
    if (!user) {
      throw new NotFoundException('User not found')
    }

    const role = await this.roleRepository.findOne({
      where: { id: addUserToOrganisationDto.roleId, organisation: { id: organisationId } },
    })
    if (!role) {
      throw new NotFoundException('Role not found in this organisation')
    }

    const existing = await this.membershipRepository.findOne({
      where: { user: { id: user.id }, organisation: { id: organisationId } },
    })
    if (existing) {
      throw new ConflictException('User is already a member of this organisation')
    }

    const membership = this.membershipRepository.create({
      user,
      role,
      organisation: { id: organisationId },
      isOwner: false,
    })

    const saved = await this.membershipRepository.save(membership)

    await this.emailService.send({
      to: user.email,
      subject: 'You were added to an organisation',
      text: `You have been added to an organisation with the role "${role.name}".`,
    })

    return saved
  }

  async update(id: number, updateOrganisationDto: UpdateOrganisationDto, user: JwtPayload) {
    const organisation = await this.ensureOwner(id, user.sub)
    Object.assign(organisation, updateOrganisationDto)
    return this.organisationRepository.save(organisation)
  }

  async remove(id: number, user: JwtPayload) {
    const organisation = await this.ensureOwner(id, user.sub)

    const invitations = await this.invitationRepository.find({ where: { organisation: { id } } })
    await this.invitationRepository.remove(invitations)

    const memberships = await this.membershipRepository.find({ where: { organisation: { id } } })
    await this.membershipRepository.remove(memberships)

    const roles = await this.roleRepository.find({ where: { organisation: { id } } })
    await this.roleRepository.remove(roles)

    return this.organisationRepository.remove(organisation)
  }

}
